src/fabric/src/fabric2_db_expiration.erl

```
@@ -0,0 +1,273 @@
+    ok = wait_for_couch_jobs_app(),
+    ok = couch_jobs:set_type_timeout(?JOB_TYPE, ?JOB_TIMEOUT_SEC),
+    ok = maybe_add_job(),
+    Pid = spawn_link(?MODULE, cleanup, [is_enabled()]),
+    {noreply, St#st{job = Pid}};
+
+handle_info({'EXIT', Pid, Exit}, #st{job = Pid} = St) ->
+    case Exit of
+        normal -> ok;
+        Error -> couch_log:error("~p : job error ~p", [?MODULE, Error])
+    end,
+    NewPid = spawn_link(?MODULE, cleanup, [is_enabled()]),
+    {noreply, St#st{job = NewPid}};
+
+handle_info(Msg, St) ->
+    {stop, {bad_info, Msg}, St}.
+
+
+code_change(_OldVsn, St, _Extra) ->
+    {ok, St}.
+
+
+wait_for_couch_jobs_app() ->
+    % Because of a circular dependency between couch_jobs and fabric apps, wait
+    % for couch_jobs to initialize before continuing. If we refactor the
+    % commits FDB utilities out we can remove this bit of code.
+    case lists:keysearch(couch_jobs, 1, application:which_applications()) of
+        {value, {couch_jobs, _, _}} ->
+            ok;
+        false ->
+            timer:sleep(100),
+            wait_for_couch_jobs_app()
+    end.
+
+
+maybe_add_job() ->
+    case couch_jobs:get_job_data(undefined, ?JOB_TYPE, job_id()) of
+        {error, not_found} ->
+            Now = erlang:system_time(second),
+            ok = couch_jobs:add(undefined, ?JOB_TYPE, job_id(), #{}, Now);
+        {ok, _JobData} ->
+            ok
+    end.
+
+
+cleanup(false) ->
+    timer:sleep(?CHECK_ENABLED_SEC * 1000),
+    exit(normal);
+
+cleanup(true) ->
+    Now = erlang:system_time(second),
+    ScheduleSec = schedule_sec(),
+    Opts = #{max_sched_time => Now + min(ScheduleSec div 3, 15)},
+    case couch_jobs:accept(?JOB_TYPE, Opts) of
+        % maybe handle timeout here, need to check the api
+        {ok, Job, Data} ->
+            try
+                {ok, Job1, Data1} = process_expirations(Job, Data),
+                ok = resubmit_job(Job1, Data1, schedule_sec())
+            catch
+                _Tag:Error ->
```

Let's add a test for this scenario. Maybe throw an error in the processing part and ensure the job still rescheduled. Since `cleanup/1` is exported we can call that directly from the eunit code and mock a failure. Then can assert that the job exits with a stack and the error and couch_jobs